tests/webTables.spec.ts

```diff
@@ -1,222 +1,133 @@
-                // Find table row "name" to validate the correct pet name appears
-                // Assumes there is only one pet in the Owner Information page
-                await expect(page.locator('dt', {hasText: 'Name'}).locator('..').locator('dd').first()).toHaveText(petName)
+test.describe('Validate Owner Information', () => {
+    
+    test.beforeEach(async ({ page }) => {
+        /* 1. Select the OWNERS menu item in the navigation bar and then select "Search" from the drop-down menu */
+        // Navigate to the owners table
+        await page.getByText('Owners').click()
+        await page.getByText('Search').click()
+    })
+    test('Validate the pet name city of the owner', async ({ page }) => {
+        /* 2. In the list of Owners, locate the owner by the name "Jeff Black". 
+        Add the assertions that this owner is from the city of "Monona" and he has a pet with a name "Lucky" */
+        // Find Jeff Black in the table 
+        const row = page.getByRole('row', {name: 'Jeff Black'})
+    
+        // Validate the owner is from Monona
+        await expect(row.locator('td').nth(2)).toHaveText('Monona')
+    
+        // Validate the owner has a pet named Lucky
+        await expect(row.locator('td').nth(4)).toHaveText('Lucky')
+    
+    });
+    
+    test('Validate owners count of the Madison city', async ({ page }) => {
+        /* 2. In the list of Owners, locate all owners who live in the city of "Madison". 
+        Add the assertion that the total number of owners should be 4 */
+        
+        // Validate there are only 4 entries of 'Madison' in the city column
+        await expect(page.getByRole('row', {name: 'Madison'})
+            // Filtering specifically for the City column
+            .filter({has: page.getByRole('cell').nth(2)})).toHaveCount(4)
+    
+    });
+    
+    test('Validate search by Last Name', async ({ page }) => {
+        /* 2. - 8. In the "Last name" input field, type the last name "Black", "Davis", "Es" (partial string), and "Playwright". */
+        // Search for the following last names and validate the results
+        const lastNames = ['Black', 'Davis', 'Es', 'Playwright']
+    
+        for(const lastName of lastNames){
+            // Use the search field to locate entries with a specific last name
+            await page.getByRole('textbox').fill(lastName)
+            await page.getByText('Find Owner').click()
+            
+            // Capture the rows in the table body (not the header)
+            const rows = await page.locator('tbody tr').all()
+    
+            // For each resulting row, check for the last name
+            for(const row of rows){
+    
+                // Ensuring the resulting row has cells. There is an extra row in the 'Pets' column that we need to filter out, as it contains no cells.
+                if((await row.getByRole('cell').all()).length > 0){
```

Comment:
Simplify!
You don't need to ensure that rows has cells. This is not needed extra step of complexity.

Reply:
Well, I threw that in there because I was having issues with the row contents.  I run this without the cell check.

![image](https://github.com/user-attachments/assets/708a27b9-6881-4120-99ae-d663122a8e6b)


And then I observed these results:
![image](https://github.com/user-attachments/assets/bd26abe6-dd84-416c-80c2-083e9b3391bd)

When I filter out the empty rows, the test passes.

Reply:
I think this is due to the embedded table rows within the cells in the Pets column.

Reply:
Okay, here is what is happening here.
Do you see in the response, that your object has all kinds of Owners? But it should have ONLY the owners that you are requesting to filter at line 49. So your loop is going through the unfiltered owners list.
Why does that happen?
Because, reminder one more time, `all()` method does not wait for the particular state of the page. The next line of code, where you get the list of row executed immediately before the table filtered.

Okay, the second problem, is that test fails because locator is incorrect.
Why? Everything seems logical. But.. look at the DOM
![image](https://github.com/user-attachments/assets/9a220b5d-5ca6-4029-8b75-4a04fae55bb2)
There is another "tr" tag inside of final column "td".
So by providing locator `page.locator('tbody tr').all()` you return all tr-s, including from the last column. So you are not looping through the list of owner rows. You are trying to loop through rows of table and row of the last column.

Need to find something more suitable... oh look, the owner name cell has a dedicated class! Can we loop through those? Absolutely!
![image](https://github.com/user-attachments/assets/e64be49f-f91e-43a7-b3e1-c647ebc1c643)

So your locator changes to this:  `page.locator('.ownerFullName').all()`
And then later in the step, you don't need to get a value of cells, because it'a already a value of cell in the row, so you just need `await expect(row).toContainText(lastName)`

Now everything magically works :)

Reply:
Thank you so much, Artem!  I'll make the necessary changes.

Reply:
This solution is awesome.  I will ask some questions in the chat.